---
title: "Sequencing Quality"
output: html_document
params:
  title: "EF01: Raw Reads from Illumina MiSeq"
  author: "Ulrich Aschl"
  ngs_run_date: "2020-06-30"
  date: !r Sys.Date()
  fwd_reads: !r c("../input_data/R0_S1_L001_R1_001.fastq", "../input_data/R0_S1_L001_R1_001.fastq")
  rev_reads: !r c("../input_data/R0_S1_L001_R2_001.fastq", "../input_data/R0_S1_L001_R2_001.fastq")
  wd: getwd()
---

# Overview
**Experiment**: `r params$title`

**NGS Run Date**: `r params$ngs_run_date`

**Analysis Date**: `r params$date`

**Author**: `r params$author`

## Overall NGS FASTQ Quality
```{r, fig.show="hold", out.width="50%", message=FALSE, warning=FALSE, echo=FALSE}
  wd <- getwd()
  setwd(params$wd)
  fwd <- dada2::plotQualityProfile(params$fwd_reads, aggregate = TRUE)
  plot(fwd)
  if (!is.null(params$rev_reads) && length(params$rev_reads) > 0) {
    rev <- dada2::plotQualityProfile(params$rev_reads, aggregate = TRUE)
    plot(rev)
  }
  
  setwd(wd)
```




## SELEX Round FASTQ Quality

Quality Plots are created using the DADA2 Pipeline's 'plotQualityProfile' function.

```{r, message=FALSE, echo=FALSE, warning=FALSE}
out = NULL
fwd_fastq <- NULL
rev_fastq <- NULL

for (i in 1:length(params$fwd_reads)) {
  fwd_fastq <- params$fwd_reads[i]
  if (!is.null(params$rev_reads) && length(params$rev_reads) > 0) {
    rev_fastq <- params$rev_reads[i]
  }
  out <- c(out, knitr::knit_child(here::here("bin", "ngs_quality_report_plot.Rmd"), quiet=TRUE))
}
```

```{r, echo=FALSE, warning=FALSE, results="asis"}
cat(paste(out, collapse = '\n'))
```
<!-- `r paste(knitr::knit(text = out, quiet=TRUE), collapse = 'asdf')` -->

